### ISSUE TITLE
Preassembler changes statements passed as argument

### ISSUE THREAD

**Issue by bgyori** [sorgerlab/indra#62]
Assume we have a processor called rp with len(rp.statements) > 0.

```
pa = Preassembler(eh, mh)
pa.add_statements(rp.statements)
```
At this point rp.statements is unchanged.

```
duplicate_stmts = pa.combine_duplicates()
```
At this point rp.statements has changed! This obviously shouldn't happen.

**Comment by johnbachman** [comment]
Aha. I wonder if this is an issue with whether the statements are copied
before manipulation. Might be fixed with a one-line copy operation.

**Comment by jmuhlich** [comment]
I think you probably just need something like this in Preassembler.add_statements:

self.stmts += copy.deepcopy(stmts)

(and import copy at the top of course)

Unless there are shared data structures within the Statements that should remain shared, this is what you want.

**Comment by bgyori** [comment]
Alright, this is an easy fix. Unfortunately many of our existing tests actually rely on two statements being the same object because of the lack of deep copying :)

**Comment by johnbachman** [comment]
That should able to be fixed through the use of the matches method, I hope?

**Comment by bgyori** [comment]
Unfortunately, no, because we need to look at the evidence when trying to establish that statements match exactly. The matches methods are (for exactly this reason) not overloading == because they ignore evidence. So now I'm implementing __eq__ and __ne__ for all of statements.py so that we can check for actual equality.

**Comment by johnbachman** [comment]
You could have eq call matches and then only inspect the Evidence objects.

**Comment by jmuhlich** [comment]
Know that once you implement \_\_eq\_\_ you'll need to implement \_\_hash\_\_ too. See here: https://docs.python.org/2/reference/datamodel.html#object.__hash__ . Instead of opening that can of worms you might want to implement another "matches" method like `matches_full` and rewrite the tests to use that. Ultimately we probably want purely immutable Statements based on tuples or something like that to make all of this much easier, but that's a more serious refactoring.

Also note that if you do choose to implement \_\_eq\_\_, your \_\_ne\_\_ implementation should be ```not self == other```. Don't do ```not self.__eq__(other)```. The == version allows a subclass to properly impose its own rules for equality if appropriate, whereas the \_\_eq\_\_ version will override it.

**Comment by bgyori** [comment]
Well, I might have just wasted a lot of time because I already implemented __eq__ and __ne__ (I implemented it the way you suggest in the first place). I thought that Statements being modified by downstream functions was the reason behind a bug but it turns out it isn't.

**Comment by bgyori** [comment]
Well, this simple issue led to a whole bunch of other changes and all should be resolved as of 145d96b4d42ed4d95db7a23ff03ab2e4e748bd8c.
